### PR TITLE
[videoplayer] Forced Subs with original stream language bug

### PR DESCRIPTION
**Pull request by onixx**
In the PredicateSubtitleFilter class, when original stream language is selected for subs, there is a selection rule that makes a sub relevant if its forced and (original is selected OR language matches).

This causes a incorrect language forced sub stream to be enabled if no forced sub stream exists in the original audio stream language.

This can happen when a file contains a forced sub track that is not required in the original audio stream language . As an example, in English, no forced subs are required if video stream has "5 years earlier" burned in video.

Without this pull request, if a french forced sub stream exists that has this sentence translated, the french forced sub is relevant and activated when playing. It should not be relevant nor enabled since the original stream language is English.

The same changes was made to the PredicateSubtitlePriority class so the french forced sub is not prioritized over the original stream language subs.

### CODE REVIEW

**Comment by MartijnKaijser** [comment]
@ace20022 ping

**Comment by ace20022** [comment]
Sounds reasonable. There were lengthy discussions back then, but I can't remember why it was done this way. I'll have a closer look next week.

**Comment by ace20022** [comment]
@onixx What are your exact settings and how does the streams look like (type, lang, flag)?

**Comment by ace20022** [comment]
For reference http://forum.kodi.tv/showthread.php?tid=176428 is the original test thread ;) I'm trying to find out why I added that original check.

**Comment by onixx** [comment]
@ace20022 I have carefully read the thread. It looks like [http://forum.kodi.tv/showthread.php?tid=176428&pid=1619774#pid1619774](url) is the user comment that made you change your mind. The problem with this logic is that it tries to compensate for users which files being played do not have language codes set properly on each streams. Which I think is an exception we should not try to compensate for.  There is certainly ways to figure if there are no language set on the sub but really, do we want to complicate this...

With the current logic, we end up with forced subs activated in a foreign language. This happens frequently since often forced subs are only required for the foreign language... not for the original language ;) (in the case I explained in my earlier post with "5 years earlier..." burned in english.

Example;
Original selected for both audio and subs.

Tracks;
audio english (default)
audio french

sub english
sub english (default)
sub french
sub french (forced)

When playing this example, audio is in english and forced french subs appear.
The problem will only occur if there are no english forced.

**Comment by ace20022** [comment]
ok, so what do you think if we change the original case to:

lang & default > default & forced (ignoring lang) > default (ignoring lang) > none

Just to be sure, in your example above no subs are displayed with this pr?

**Comment by onixx** [comment]
do you mean
lang & forced > default & forced (ignoring lang) > default (ignoring lang) > none

**Comment by ace20022** [comment]
yes, the last one missing is lang & default, right? Is this second best?

**Comment by onixx** [comment]
Hello, I am doing some more tests. stay tuned

**Comment by onixx** [comment]
Ok, this is the revisions;

For PredicateSubtitleFilter, 
flag forced matching language returns false
flag forced and flag default returns false
(removed flag default alone returning false and )

For PredicateSubtitlePriority
1. forced in right language
2. default in right language
3. forced and default whatever the language

On my end, I have the expected results. Bottom line, we no longer turn on forced subs when original stream language is selected unless language matches OR stream is (forced and default)

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp [hunk 1]
```diff
@@ -239,7 +238,9 @@ static bool PredicateAudioPriority(const SelectionStream& lh, const SelectionStr
 *          A subtitle lh is 'better than' a subtitle rh (in evaluation order) if
 *          - lh was previously selected, or
 *          - lh is an external sub and rh not, or
-*          - lh is a forced sub and ("original stream's language" was selected or subtitles are off) and rh not, or
+*          - lh is a forced sub and ("original stream's language" was selected or subtitles are off) and audio stream language matches sub language and rh not, or
+*          - lh is a default sub and ("original stream's language" was selected or subtitles are off) and audio stream language matches sub language and rh not, or
+*	   - lh is a forced sub and a default sub ("original stream's language" was selected or subtitles are off)
```
Comment (ace20022) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp]: indentation

**Comment by ace20022** [comment]
@onixx sorry for the slow respond speed, I'm quite busy at the moment.
So ```default``` is only meaningful when it matches the language? Imo if nothing matches then a default sub should be selected, no?

**Comment by onixx** [comment]
@ace20022 no problem. going 24/7 here ! baby arrived last week so no sleep.

default alone no longer triggers sub to be turned on in PredicateSubtitleFilter. The stream has to be forced in right language or it has to be flagged default and forced as far as the changes I made. Default alone does not turn them on.

However, as far as priority, default still relevant in PredicateSubtitlePriority, very last test (left as is)

Also, after doing some more tests, I realized that there was no ranking in PredicateSubtitlePriority for !subson || original with streams that did not bear forced or default but are in original stream language. So now, the priority is:
forced in right language, 
default in right language, 
just in right language, 
whatever language forced and default... 
and finally, default only as the last check in PredicateSubtitlePriority.

I have not yet committed this last change. Testing some more

Thanks !

**Comment by onixx** [comment]
@ace20022 I have tested a number of scenarios with different combination of language, default and forced flags and it all works great.

In summary these are the changes:

---- with original stream language selected:

-Sub track is automatically selected (but not necessarily turned on) in this order:
forced in right language, default in right language, just in right language, whatever language forced and default, default only

-Sub track is automatically turned on when (forced flag and language match) or (forced and default)

**Comment by ace20022** [comment]
Great! Could you please squash and rebase on top of our master branch? We don't want that merge... commits.

**Comment by onixx** [comment]
@ace20022 done. Thanks !

**Comment by ace20022** [comment]
jenkins build this please

**Pull request merged**